--- a/sources/mma/bcknd/cpu/mma_cpu.f90
+++ b/sources/mma/bcknd/cpu/mma_cpu.f90
@@ -320,6 +320,19 @@
          pij => this%pij%x, qij => this%qij%x, &
          low => this%low%x, upp => this%upp%x)
 
+#ifdef AAGE_REG
+      ! topopt_in_petsc MMA::GenSub, constraintModification = PETSC_FALSE
+      p0j = (upp - x)**2 * (max(df0dx, 0.0_rp) + 0.001_rp*abs(df0dx) &
+           + 0.5_rp*1.0e-6_rp/(upp - low))
+      q0j = (x - low)**2 * (max(-df0dx, 0.0_rp) + 0.001_rp*abs(df0dx) &
+           + 0.5_rp*1.0e-6_rp/(upp - low))
+      do j = 1, this%n
+         do i = 1, this%m
+            pij(i, j) = (upp(j) - x(j))**2 * max(dfdx(i, j), 0.0_rp)
+            qij(i, j) = (x(j) - low(j))**2 * max(-dfdx(i, j), 0.0_rp)
+         end do
+      end do
+#else
       p0j = ( &
            1.001_rp * max(df0dx, 0.0_rp) &
            + 0.001_rp * max(-df0dx, 0.0_rp) &
@@ -348,6 +361,7 @@
          end do
       end do
 
+#endif
     end associate
 
     ! ------------------------------------------------------------------------ !
@@ -835,7 +849,7 @@
     integer, dimension(this%m+1) :: ipiv
 
     ! Parameters for global communication
-    real(kind=rp) :: minimal_epsilon
+    real(kind=rp) :: minimal_epsilon, err_stale
 
     ! ------------------------------------------------------------------------ !
     ! initial value for the parameters in the subsolve based on
@@ -853,7 +867,12 @@
     ! ------------------------------------------------------------------------ !
     ! Computing the minimal epsilon and choose the most conservative one
 
+#ifdef AAGE_TERM
+    minimal_epsilon = this%epsimin
+    err_stale = 1.0_rp
+#else
     minimal_epsilon = max(0.9_rp * this%epsimin, 1.0e-12_rp)
+#endif
     call MPI_Allreduce(MPI_IN_PLACE, minimal_epsilon, 1, &
          mpi_real_precision, MPI_MIN, neko_comm, ierr)
 
@@ -917,11 +936,18 @@
               mpi_real_precision, MPI_MAX, neko_comm, ierr)
 
          ! ------------------------------------------------------------------- !
+#ifdef AAGE_TERM
+         residual_max = err_stale
+#endif
          ! Internal loop
          do iter = 1, this%max_iter
 
             !Check the condition
+#ifdef AAGE_TERM
+            if (.not. (residual_max .gt. 0.9_rp * epsi)) exit
+#else
             if (residual_max .lt. epsi) exit
+#endif
 
             ! Compute dL(x, y, z, λ)/dλ for the updated x(λ), y(λ), z(λ)
 
@@ -1069,6 +1095,9 @@
             residual_max = maxval(abs([relambda, remu]))
             call MPI_Allreduce(MPI_IN_PLACE, residual_max, 1, &
                  mpi_real_precision, MPI_MAX, neko_comm, ierr)
+#ifdef AAGE_TERM
+            err_stale = residual_max
+#endif
          end do
        end associate
 
